zio/zjsonio/reader.go

```diff
@@ -166,31 +167,63 @@ func decodeType(columns []interface{}) (string, error) {
 }
 
 func decodeField(builder *zcode.Builder, typ zng.Type, s string) error {
-	var err error
-	var index int
 	b := []byte(s)
 	if zng.IsContainerType(typ) && !zng.IsUnionType(typ) {
 		return zng.ErrNotContainer
 	}
-	if utyp, ok := typ.(*zng.TypeUnion); ok {
-		typ, index, b, err = utyp.SplitZng(b)
-		if err != nil {
-			return err
-		}
-		builder.BeginContainer()
-		defer builder.EndContainer()
-		var a [8]byte
-		n := zcode.EncodeCountedUvarint(a[:], uint64(index))
-		builder.AppendPrimitive(a[:n])
-	}
-
 	zv, err := typ.Parse(b)
 	if err != nil {
 		return err
 	}
 	builder.AppendPrimitive(zv)
 	return nil
+}
 
+func decodeUnion(builder *zcode.Builder, typ *zng.TypeUnion, body interface{}) error {
+	builder.BeginContainer()
+	tuple, ok := body.([]interface{})
+	if !ok || len(tuple) != 2 {
+		return errors.New("bad json for zjson union value")
+	}
+	istr, ok := tuple[0].(string)
+	if !ok {
+		return errors.New("bad type index for zjson union value ")
+	}
+	index, err := strconv.Atoi(istr)
+	if err != nil {
+		return fmt.Errorf("bad type index for zjson union value: %w", err)
+	}
```

Comment:
Nit: Encoding the index as a JSON number instead of a string seems like it would be more natural.

Reply:
JSON numbers are floats and so it didn't seem particularly natural to use a json number for an integer index here. Also everything a zjson "value" is encoded as a string, including ints.

Since you prefixed this with "nit" I'm assuming you're ok with the current scheme, but I can change  to numbers if you'd like.